### PR TITLE
DOC: Adapt enum section to strongly typed enums

### LINKED ISSUE #154: Describe the use of strongly typed enums
Following [ITK issue 1049](https://github.com/InsightSoftwareConsortium/ITK/issues/1049) and the corresponding PRs, ITK uses strongly typed enums. The ITK v5 migration guide provides details about the current practice in ITK:
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Documentation/ITK5MigrationGuide.md#strongly-typed-enumerations

This practice should be reflected by the ITK SW guide, including how printing those types is being tested and the potential changes required in `CMakeLists.txt` files. Section `Naming Enums` (section C.6.8, p.321) is a possible candidate to cover the topic.

A quick search shows that the file `ModelBasedSegmentation.cxx` makes use of enums, so it will require adapting it to modern ITK practice.

### PR DESCRIPTION
**Pull request by jhlegarreta**
Adapt section about enums to the adoption of strongly typed enums in ITK:
https://github.com/InsightSoftwareConsortium/ITK/commit/0dd31a62f851a0ae95ef4c6d6c85e531a4ebca59

Fixes #154 .

### CODE REVIEW

File: SoftwareGuide/Latex/Appendices/CodingStyleGuide.tex [hunk 1]
```diff
@@ -758,50 +758,112 @@ \subsection{Naming Class Data Members}
...
-added for each enum-list entry, if necessary, need not to be aligned.
+ITK uses strongly-typed enumerations through the \code{enum class} keyword.
+Enumerations in ITK need to declared within a dedicated class and be declared
+as \code{public}. The class name must be append appended with the \code{Enum}
+(or \code{Enums} if the class contains several strongly-typed enumerations)
+label. Enumeration classes must declare their identifier before the enum-list
+is specified, it must start with capitals and be written with case change, and
+it shall not bear the \code{Type} appendix. The type of the enumeration must be
+defined explicitly. The streaming operator \code{<<} must be overloaded to
+define enumerations are printed. The enum-list must be specified in capitals.
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/Appendices/CodingStyleGuide.tex:772]: >  The enum-list must be specified in capitals.

@hjmjohnson Not sure if this is still preferred.

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex [hunk 2]
```diff
@@ -739,6 +739,25 @@ \subsubsection{Wrapping Variables}
...
+itk_wrap_simple_class("itk::MathematicalMorphologyEnums")
+\end{minted}
+\normalsize
+
+which results in access to the class in Python as
+\code{itk.MathematicalMorphologyEnum}, its \code{Algorithm} enumeration class
+being accesible as \code{itk.MathematicalMorphologyEnums.Algorithm}, and its
+enum-list being accessed as
+\code{itk.MathematicalMorphologyEnums.Algorithm.BASIC},
+\code{itk.MathematicalMorphologyEnums.Algorithm.HISTO}, etc.
```
Comment (jhlegarreta) [inline SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:758]: @thewtex Have not tested any of these cases.

**Comment by jhlegarreta** [comment]
I think it is OK that the enum in `ModelBasedSegmentation.cxx`:
https://github.com/InsightSoftwareConsortium/ITKSoftwareGuide/blob/master/SoftwareGuide/Cover/Source/ModelBasedSegmentation.cxx#L131

continues to be an unscoped enumeration. @thewtex @hjmjohnson let me know if it should be transitioned to a strongly-typed enumeration.

File: SoftwareGuide/Latex/Appendices/CodingStyleGuide.tex [hunk 3]
```diff
@@ -758,50 +758,112 @@ \subsection{Naming Class Data Members}
...
 \label{subsec:NamingEnums}
 
-Enumeration list (\code{enum}) must declare its identifier before the
-enum-list is specified, it must start with capitals and be written with case
-change, it should generally add the \code{Type} appendix (e.g. \code{MyEnumType}),
-and the enum-list must be specified in capitals. The documentation or comments
-added for each enum-list entry, if necessary, need not to be aligned.
+ITK uses strongly-typed enumerations through the \code{enum class} keyword.
+Enumerations in ITK need to declared within a dedicated class and be declared
+as \code{public}. The class name must be append appended with the \code{Enum}
```
Comment (thewtex) [inline SoftwareGuide/Latex/Appendices/CodingStyleGuide.tex:766]: extra *append*

File: SoftwareGuide/Latex/Appendices/CodingStyleGuide.tex [hunk 4]
```diff
@@ -758,50 +758,112 @@ \subsection{Naming Class Data Members}
...
+(or \code{Enums} if the class contains several strongly-typed enumerations)
```
Comment (thewtex) [inline SoftwareGuide/Latex/Appendices/CodingStyleGuide.tex:767]: Always *Enums*?

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex (hunk 2 above)
Comment (thewtex) [reply SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex:758]: @jhlegarreta close, `Algorithm.BASIC` -> `Algorithm_BASIC`, etc.

**Review by thewtex — CHANGES_REQUESTED** [review]
@jhlegarreta thank you so much for writing this up!!

A few items noted inline

File: SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex (hunk 2 above)
Comment (jhlegarreta) [reply SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex]: Changed in 088a010.

File: SoftwareGuide/Latex/Appendices/CodingStyleGuide.tex (hunk 3 above)
Comment (jhlegarreta) [reply SoftwareGuide/Latex/Appendices/CodingStyleGuide.tex]: Fixed in 088a010.

File: SoftwareGuide/Latex/Appendices/CodingStyleGuide.tex (hunk 4 above)
Comment (jhlegarreta) [reply SoftwareGuide/Latex/Appendices/CodingStyleGuide.tex]: :+1: Changed in 088a010.

**Comment by jhlegarreta** [comment]
Thanks for the review @thewtex .

**Review by thewtex — APPROVED** [review]
🏅 👏 🙏

**Pull request merged**